presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java

```diff
@@ -78,118 +80,151 @@ public OperatorFactory duplicate()
 
     @Override
-    public boolean isFinished()
+    public boolean needsInput()
     {
-        return finishing && outputPage == null;
+        return pageBuffer.isEmpty() && !pageBuffer.isFinished();
     }
 
     @Override
-    public ListenableFuture<?> isBlocked()
+    public void addInput(Page page)
     {
-        return channelSetFuture;
+        pageBuffer.add(page);
     }
 
     @Override
-    public boolean needsInput()
+    public WorkProcessor<Page> getOutputPages()
     {
-        if (finishing || outputPage != null) {
-            return false;
-        }
-
-        if (channelSet == null) {
-            channelSet = tryGetFutureValue(channelSetFuture).orElse(null);
-        }
-        return channelSet != null;
+        return pages;
     }
 
     @Override
-    public void addInput(Page page)
+    public void close()
+            throws Exception
     {
-        requireNonNull(page, "page is null");
-        checkState(!finishing, "Operator is finishing");
-        checkState(channelSet != null, "Set has not been built yet");
-        checkState(outputPage == null, "Operator still has pending output");
-
-        // create the block builder for the new boolean column
-        // we know the exact size required for the block
-        BlockBuilder blockBuilder = BOOLEAN.createFixedSizeBlockBuilder(page.getPositionCount());
-
-        Page probeJoinPage = new Page(page.getBlock(probeJoinChannel));
-        Optional<Block> hashBlock = probeHashChannel.map(page::getBlock);
-
-        // update hashing strategy to use probe cursor
-        for (int position = 0; position < page.getPositionCount(); position++) {
-            if (probeJoinPage.getBlock(0).isNull(position)) {
-                if (channelSet.isEmpty()) {
-                    BOOLEAN.writeBoolean(blockBuilder, false);
-                }
-                else {
-                    blockBuilder.appendNull();
-                }
+    }
+
+    private class SemiJoinPages
+            implements WorkProcessor.Transformation<Page, Page>
+    {
+        private final int probeJoinChannel;
+        private final ListenableFuture<ChannelSet> channelSetFuture;
+        private final Optional<Integer> probeHashChannel;
+
+        private ChannelSet channelSet;
+
+        public SemiJoinPages(SetSupplier channelSetFuture, int probeJoinChannel, Optional<Integer> probeHashChannel)
+        {
+            requireNonNull(channelSetFuture, "hashProvider is null");
+            checkArgument(probeJoinChannel >= 0, "probeJoinChannel is negative");
+
+            this.channelSetFuture = channelSetFuture.getChannelSet();
+            this.probeJoinChannel = probeJoinChannel;
+            this.probeHashChannel = probeHashChannel;
+        }
+
+        @Override
+        public WorkProcessor.TransformationState<Page> process(Page inputPage)
+        {
+            if (!this.channelSetFuture.isDone()) {
+                return WorkProcessor.TransformationState.blocked(this.channelSetFuture);
```

Comment:
technically, single input page will be buffered until future is done. We could account it's memory here